src/main/scala/com/virtuslab/shared_indexes/remote/S3Upload.scala

```diff
@@ -0,0 +1,128 @@
+    if (failedRemoves.nonEmpty) {
+      failureSummary.append(s"Failed to remove ${failedRemoves.size} entries: \n")
+      failedRemoves.foreach { case (entry, e) =>
+        failureSummary.append(s"  $entry: \n    ${e.getMessage}\n")
+      }
+    }
+    if (failedUploads.nonEmpty) {
+      failureSummary.append(s"Failed to upload ${failedUploads.size} entries: \n")
+      failedUploads.foreach { case (entry, e) =>
+        failureSummary.append(s"  $entry: \n    ${e.getMessage}\n")
+      }
+    }
+    if (failureSummary.nonEmpty) {
+      throw new RuntimeException(failureSummary.toString())
+    }
+  }
+
+  private def uploadFiles(files: Iterable[CdnUploadEntry]): Map[CdnUploadEntry, Throwable] = {
+    logger.info("Uploading {} items...", files.size)
+    val results = awaitAll(putObject)(files)
+    collectFailures(files, results)
+  }
+
+  private def removeFiles(files: Iterable[CdnEntry]): Map[CdnEntry, Throwable] = {
+    logger.info("Removing {} items...", files.size)
+    val results = awaitAll(deleteObject)(files)
+    collectFailures(files, results)
+  }
+
+  private def awaitAll[T, R](task: T => Future[R])(inputs: Iterable[T]): Iterable[Try[R]] = {
+    val futures = inputs.map(task)
+    val completions = futures.map(_.transform { case _: Try[R] => Success(()) })
+    val allCompleted = Future.sequence(completions)
+    var timeoutException: Option[TimeoutException] = None
+    try {
+      Await.ready(allCompleted, timeout)
+    } catch {
+      case e: TimeoutException => timeoutException = Some(e)
+    }
+    futures.map(_.value.getOrElse(Failure(timeoutException.get)))
+  }
+
+  private def collectFailures[T, R](keys: Iterable[T], results: Iterable[Try[R]]): Map[T, Throwable] = {
+    (keys zip results).collect { case (file, Failure(e)) =>
+      (file, e)
+    }.toMap
+  }
+
+  private def putObject(entry: CdnUploadEntry): Future[Unit] = {
+    require(
+      !entry.getKey.startsWith(s"${rootInBucket}/"),
+      s"Key must not start with S3 paths prefix (${rootInBucket}): ${entry.getKey}"
+    )
+    val req = PutObjectRequest.builder()
+      .bucket(s3.getBucket)
+      .key(keyInBucket(entry))
+      .contentType(entry.getContentType)
+      .build()
+    Future {
+      logger.info("S3PUT {}", entry)
+      Using.resource(getInputStream(entry)) { istream =>
+        s3.getClient.putObject(req, RequestBody.fromInputStream(istream, entry.getContentLength))
+      }
+      logger.info("S3PUT end {}", entry)
+    }
+  }
+
+  private def deleteObject(entry: CdnEntry): Future[Unit] = {
+    require(
+      !entry.getKey.startsWith(s"${rootInBucket}/"),
+      s"Key must not start with S3 paths prefix (${rootInBucket}): ${entry.getKey}"
+    )
+    require(entry.isInstanceOf[S3CdnEntry], s"It is only possible to remove S3 entries, but was: $entry")
+    val req = DeleteObjectRequest.builder()
+      .bucket(s3.getBucket)
+      .key(keyInBucket(entry))
+      .build()
+    Future {
+      logger.info("S3DEL {}", entry)
```

Comment:
there is logger.info end for put but not for delete, why?

Reply:
I accidentally left that out after debugging. The original code doesn't log request completion.